go/vt/vtadmin/cache/cache.go

```diff
@@ -0,0 +1,262 @@
+func New[Key Keyer, Value any](fillFunc func(ctx context.Context, req Key) (Value, error), cfg Config) *Cache[Key, Value] {
+	if cfg.BackfillEnqueueWaitTime <= 0 {
+		log.Warningf("BackfillEnqueueWaitTime (%v) must be positive, defaulting to %v", cfg.BackfillEnqueueWaitTime, DefaultBackfillEnqueueWaitTime)
+		cfg.BackfillEnqueueWaitTime = DefaultBackfillEnqueueWaitTime
+	}
+
+	if cfg.BackfillRequestTTL <= 0 {
+		log.Warningf("BackfillRequestTTL (%v) must be positive, defaulting to %v", cfg.BackfillRequestTTL, DefaultBackfillRequestTTL)
+		cfg.BackfillRequestTTL = DefaultBackfillRequestTTL
+	}
+
+	c := &Cache[Key, Value]{
+		cache:     cache.New(cfg.DefaultExpiration, cfg.CleanupInterval),
+		fillcache: cache.New(cfg.BackfillRequestDuplicateInterval, time.Minute),
+		backfills: make(chan *backfillRequest[Key], cfg.BackfillQueueSize),
+		fillFunc:  fillFunc,
+		cfg:       cfg,
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+
+	c.wg.Add(1)
+	go c.backfill() // TODO: consider allowing N backfill threads to run, configurable
+
+	return c
+}
+
+// Add adds a (key, value) to the cache directly, following the semantics of
+// (github.com/patrickmn/go-cache).Cache.Add.
+func (c *Cache[Key, Value]) Add(key Key, val Value, d time.Duration) error {
+	return c.add(key.Key(), val, d)
+}
+
+func (c *Cache[Key, Value]) add(key string, val Value, d time.Duration) error {
+	// Record the time we last cached this key, to check against
+	c.fillcache.Set(key, struct{}{}, cache.DefaultExpiration)
+	// Then cache the actual value.
+	return c.cache.Add(key, val, d)
+}
+
+// Get returns the Value stored for the key, if present in the cache. If the key
+// is not cached, the zero value for the given type is returned, along with a
+// boolean to indicated presence/absence.
+func (c *Cache[Key, Value]) Get(key Key) (Value, bool) {
+	v, exp, ok := c.cache.GetWithExpiration(key.Key())
+	if !ok || (!exp.IsZero() && exp.Before(time.Now())) {
+		var zero Value
+		return zero, false
+	}
+
+	return v.(Value), ok
+}
+
+// EnqueueBackfill submits a request to the backfill queue.
+func (c *Cache[Key, Value]) EnqueueBackfill(k Key) bool {
+	req := &backfillRequest[Key]{
+		k:           k,
+		requestedAt: time.Now().UTC(),
+	}
+
+	ctx, cancel := context.WithTimeout(c.ctx, c.cfg.BackfillEnqueueWaitTime)
+	defer cancel()
+
+	select {
+	case c.backfills <- req:
+		return true
+	case <-ctx.Done():
+		return false
+	}
+}
+
+// Close closes the backfill goroutine, effectively rendering this cache
+// unusable for further background fills.
+func (c *Cache[Key, Value]) Close() error {
+	c.cancel()
+	c.wg.Wait()
+	return nil
+}
+
+type backfillRequest[Key Keyer] struct {
+	k           Key
+	requestedAt time.Time
+}
+
+func (c *Cache[Key, Value]) backfill() {
+	defer c.wg.Done()
+
+	for {
+		var req *backfillRequest[Key]
+		select {
+		case <-c.ctx.Done():
+			return
+		case req = <-c.backfills:
+			select {
+			case <-c.ctx.Done():
+				return
+			default:
+			}
```

Comment:
This is not doing anything really -- the `ctx.Done()` could be closed throughout any of the lines of this function, so the whole function should be able to operate with a done context. Checking it just one extra time at the beginning is no different than the check in the parent `select` loop.

Reply:
Ah true, good catch!! Ideally, the given `fillFunc` should fail on a done context as well, so we should get an error back from there and then the outer `select` will pick up the done-ness on the next iteration.